lib/benchpark/cuda.py

```diff
@@ -0,0 +1,131 @@
+# Benchpark Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Dict
+import yaml  # TODO: some way to ensure yaml available
+
+from benchpark.directives import ExperimentSystemBase
+import benchpark.spec
+import benchpark.paths
+import benchpark.repo
+import benchpark.runtime
+import benchpark.variant
+
+bootstrapper = benchpark.runtime.RuntimeResources(benchpark.paths.benchpark_home)
+bootstrapper.bootstrap()
+
+import ramble.language.language_base  # noqa
+import ramble.language.language_helpers  # noqa
+
+
+class CudaExperiment(ExperimentSystemBase):
+    """This is the superclass for all benchpark experiments.
+
+    ***The Experiment class***
+
+    Experiments are written in pure Python.
+
+    There are two main parts of a Benchpark experiment:
+
+      1. **The experiment class**.  Classes contain ``directives``, which are
+         special functions, that add metadata (variants) to packages (see
+         ``directives.py``).
+
+      2. **Experiment instances**. Once instantiated, an experiment is
+         essentially a collection of files defining an experiment in a
+         Ramble workspace.
+    """
+
+    variant("cuda", default=False, description="Build and run with CUDA")
+
+
+
+    #
+    # These are default values for instance variables.
+    #
+
+    # This allows analysis tools to correctly interpret the class attributes.
+    variants: Dict[
+        "benchpark.spec.Spec",
+        Dict[str, benchpark.variant.Variant],
+    ]
+
+    def __init__(self, spec):
+        self.spec: "benchpark.spec.ConcreteExperimentSpec" = spec
+        super().__init__()
+
+    def compute_include_section(self):
+        # include the config directory
+        # TODO: does this need to change to interop with System class
+        return ["./configs"]
+
+    def compute_config_section(self):
+        # default configs for all experiments
+        return {
+            "deprecated": True,
+            "spack_flags": {"install": "--add --keep-stage", "concretize": "-U -f"},
+        }
+
+    def compute_modifiers_section(self):
+        # by default we use the allocation modifier and no others
+        return [{"name": "allocation"}]
+
+    def compute_applications_section(self):
+        # TODO: is there some reasonable default?
+        variables = {}
+        variables["n_gpus"] = num_procs  # TODO: num_procs will be defined in the child...
+
+        raise NotImplementedError(
+            "Each experiment must implement compute_applications_section"
+        )
+
+    def compute_spack_section(self):
+        # TODO: is there some reasonable default based on known variable names?
+        system_specs = {}
+        system_specs["cuda_version"] = "{default_cuda_version}"
+        system_specs["cuda_arch"] = "{cuda_arch}"
+        system_specs["blas"] = "cublas-cuda"  #TODO: can we define this in Lapack/Blas CudaExperiment?
```

Comment:
In my experience this differs from system to system, and it's best handled by saying here just that you need `blas`; the system can then register whatever `blas` external is appropriate (see for example in Sierra definition: https://github.com/LLNL/benchpark/blob/develop/var/sys_repo/systems/sierra/externals/cuda/00-version-10-1-243-packages.yaml).